src/core/context/utils.ts

```diff
@@ -206,7 +206,295 @@ export function filterMessagesByLLMCapabilities(
+        logger.warn(`Failed to filter messages by LLM capabilities: ${String(error)}`);
         return messages;
     }
 }
+
+/**
+ * Detect if a string is likely a Base64 blob (not a typical sentence/text).
+ * Uses a length threshold and character set heuristic.
+ */
+export function isLikelyBase64String(
+    value: string,
+    minLength: number = MIN_BASE64_HEURISTIC_LENGTH
+): boolean {
+    if (!value || value.length < minLength) return false;
+    // Fast-path for data URIs which embed base64
+    if (value.startsWith('data:') && value.includes(';base64,')) return true;
+    // Heuristic: base64 characters only and length divisible by 4 (allow small remainder due to padding)
+    const b64Regex = /^[A-Za-z0-9+/=\r\n]+$/;
+    if (!b64Regex.test(value)) return false;
+    // Low whitespace / punctuation typical for base64
+    const nonWordRatio = (value.match(/[^A-Za-z0-9+/=]/g)?.length || 0) / value.length;
+    return nonWordRatio < 0.01;
+}
+
+/**
+ * Parse data URI and return { mediaType, base64 } or null if not a data URI.
+ */
+export function parseDataUri(value: string): { mediaType: string; base64: string } | null {
+    if (!value.startsWith('data:')) return null;
+    const commaIdx = value.indexOf(',');
+    if (commaIdx === -1) return null;
+    const meta = value.slice(5, commaIdx); // skip 'data:'
+    if (!/;base64$/i.test(meta)) return null;
+    const mediaType = meta.replace(/;base64$/i, '') || 'application/octet-stream';
+    const base64 = value.slice(commaIdx + 1);
+    return { mediaType, base64 };
+}
+
+/**
+ * Recursively sanitize objects by replacing suspiciously-large base64 strings
+ * with placeholders to avoid blowing up the context window.
+ */
+function sanitizeDeepObject(obj: unknown): unknown {
+    if (obj == null) return obj;
+    if (typeof obj === 'string') {
+        if (isLikelyBase64String(obj)) {
+            // Replace with short placeholder; do not keep raw data
+            const approxBytes = Math.floor((obj.length * 3) / 4);
+            logger.debug(
+                `sanitizeDeepObject: replaced large base64 string (~${approxBytes} bytes) with placeholder`
+            );
+            return `[binary data omitted ~${approxBytes} bytes]`;
+        }
+        return obj;
+    }
+    if (Array.isArray(obj)) return obj.map((x) => sanitizeDeepObject(x));
+    if (typeof obj === 'object') {
+        const out: Record<string, any> = {};
+        for (const [k, v] of Object.entries(obj as Record<string, unknown>)) {
+            out[k] = sanitizeDeepObject(v);
+        }
+        return out;
+    }
+    return obj;
+}
+
+/**
+ * Convert an arbitrary tool result into safe InternalMessage content.
+ * - Converts data URIs and base64 blobs to media/file parts
+ * - Removes huge binary blobs inside objects
+ * - Truncates extremely long raw text
+ */
+export function sanitizeToolResultToContent(result: unknown): InternalMessage['content'] {
```

Comment:
add unit tests for this function for known result types to validate it works properly